python/ray/dashboard/memory.py

```diff
@@ -0,0 +1,280 @@
+            object_ref.get("submittedTaskRefCount", 0))
+        self.contained_in_owned = [
+            ray.ObjectID(decode_object_id_if_needed(object_id))
+            for object_id in object_ref.get("containedInOwned", [])
+        ]
+        self.reference_type = self._get_reference_type()
+
+    def is_valid(self) -> bool:
+        if (not self.pinned_in_memory and self.local_ref_count == 0
+                and self.submittedd_task_ref_count == 0
+                and len(self.contained_in_owned) == 0):
+            return False
+        elif self.object_id.is_nil():
+            return False
+        else:
+            return True
+
+    def group_key(self, group_by_type: GroupByType):
+        if group_by_type == GroupByType.NODE_ADDRESS:
+            return self.node_address
+        else:
+            raise ValueError(
+                "group by type {} is invalid.".format(group_by_type))
+
+    def _get_reference_type(self) -> str:
+        if self._is_object_id_actor_handle():
+            return ReferenceType.ACTOR_HANDLE
+        if self.pinned_in_memory:
+            return ReferenceType.PINNED_IN_MEMORY
+        elif self.submittedd_task_ref_count > 0:
+            return ReferenceType.USED_BY_PENDING_TASK
+        elif self.local_ref_count > 0:
+            return ReferenceType.LOCAL_REFERENCE
+        elif len(self.contained_in_owned) > 0:
+            return ReferenceType.CAPTURED_IN_OBJECT
+        else:
+            return ReferenceType.UNKNOWN_STATUS
+
+    def _is_object_id_actor_handle(self) -> bool:
+        object_id_hex = self.object_id.hex()
+        random_bits = object_id_hex[:16]
+        actor_random_bits = object_id_hex[16:24]
+
+        # random (8B) | ActorID(6B) | flag (2B) | index (6B)
+        # ActorID(6B) == ActorRandomByte(4B) + JobID(2B)
+        # If random bytes are all 'f', but ActorRandomBytes
+        # are not all 'f', that means it is an actor creation
+        # task, which is an actor handle.
+        if (random_bits == "f" * 16 and not actor_random_bits == "f" * 8):
+            return True
+        else:
+            return False
+
+    def __dict__(self):
+        return {
+            "object_id": self.object_id.hex(),
+            "pid": self.pid,
+            "node_ip_address": self.node_address,
+            "object_size": self.object_size,
+            "reference_type": self.reference_type,
+            "call_site": self.call_site,
+            "local_ref_count": self.local_ref_count,
+            "pinned_in_memory": self.pinned_in_memory,
+            "submittedd_task_ref_count": self.submittedd_task_ref_count,
+            "contained_in_owned": [
+                object_id.hex() for object_id in self.contained_in_owned
+            ],
+            "type": "Driver" if self.is_driver else "Worker"
+        }
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        return str(self.__dict__())
+
+
+class MemoryTableSummary:
```

Comment:
I think this class should be a function that returns a dictionary. In my opinion, incrementing the fields in summarize is a more verbose approach than if the entire class were replaced with a dictionary that returns the result that __dict__ currently returns. At some point, we might need the additional abstraction of this class, but I don't think we need it now.